application-xpoll-api/src/main/java/com/xwiki/xpoll/internal/DefaultXPollManager.java

```diff
@@ -0,0 +1,217 @@
+            stringBuilder
+                .append(URLEncoder.encode(documentReference.getWikiReference().getName(), XWiki.DEFAULT_ENCODING));
+            for (SpaceReference spaceReference : documentReference.getSpaceReferences()) {
+                stringBuilder.append("/spaces/");
+                stringBuilder.append(URLEncoder.encode(spaceReference.getName(), XWiki.DEFAULT_ENCODING));
+            }
+            stringBuilder.append("/pages/");
+            stringBuilder.append(URLEncoder.encode(documentReference.getName(), XWiki.DEFAULT_ENCODING));
+            stringBuilder.append("/xpoll");
+            return stringBuilder.toString();
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(String.format("Failed to retrieve the REST URL of the document: [%s]",
+                documentReference), e);
+        }
+    }
+
+    @Override
+    public Map<String, Integer> getVoteResults(DocumentReference documentReference) throws XPollException
+    {
+        XWikiContext context = contextProvider.get();
+        try {
+            XWikiDocument doc = context.getWiki().getDocument(documentReference, context);
+            List<BaseObject> xpollVotes = doc.getXObjects(XPOLL_VOTES_CLASS_REFERENCE);
+            BaseObject xpollObj = doc.getXObject(XPOLL_CLASS_REFERENCE);
+            if (xpollObj == null) {
+                throw new XPollException(String.format("The document [%s] does not have a poll object.",
+                    documentReference));
+            }
+            List<String> proposals = xpollObj.getListValue(PROPOSALS);
+            return getXPollResults(xpollVotes, proposals);
+        } catch (XWikiException e) {
+            throw new XPollException(String.format("Failed to retrieve the vote results for poll [%s]. Root cause: "
+                    + "[%s].", documentReference,
+                ExceptionUtils.getRootCauseMessage(e)));
+        }
+    }
+
+    private void setUserVotes(List<String> votedProposals, XWikiContext context, XWikiDocument doc,
+        DocumentReference user) throws XWikiException
+    {
+        String currentUserName = serializer.serialize(user, doc.getDocumentReference().getWikiReference().getName());
+        BaseObject xpollVoteOfCUrrentUser = doc.getXObject(XPOLL_VOTES_CLASS_REFERENCE, USER,
+            currentUserName, false);
+
+        if (xpollVoteOfCUrrentUser == null) {
+            xpollVoteOfCUrrentUser = doc.newXObject(XPOLL_VOTES_CLASS_REFERENCE, context);
+        }
+
+        xpollVoteOfCUrrentUser.set(USER, currentUserName, context);
+        xpollVoteOfCUrrentUser.set(VOTES, votedProposals, context);
+    }
+
+    private void updateWinner(XWikiContext context, XWikiDocument doc) throws XWikiException, XPollException
+    {
+        BaseObject xpollObj = doc.getXObject(XPOLL_CLASS_REFERENCE);
+        if (xpollObj == null) {
+            throw new XPollException(String.format("The document [%s] does not have a poll object.",
```

Comment:
This string appears twice in this file. Are you sure Checkstyle doesn't complain?